drivers/serial/uart_mcux_flexcomm.c

```diff
@@ -342,6 +379,640 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
+	} else if (channel >= 16 && channel <= 23) {
+		chmux_sel = (chmux_sel_id + 4 * (channel - 16))
+				<< CHMUX_OFF_SHIFT;
+	} else {
+		chmux_sel = (chmux_sel_id + 4 * (channel - 20))
+				<< CHMUX_OFF_SHIFT;
+	}
+
+#endif /* RT5xx */
+
+	uint32_t req_en_id = 0;
+
+	if (base == (DMA_Type *)DMA0_BASE) {
+		req_en_id = DMA0_REQ_ENA0_ID;
+	} else if (base == (DMA_Type *)DMA1_BASE) {
+		req_en_id = DMA1_REQ_ENA0_ID;
+	}
+
+
+	uint32_t en_val;
+
+	if (channel <= 31) {
+		en_val = channel + (req_en_id << ENA_SHIFT);
+	} else {
+		en_val = (channel - 32) + ((req_en_id + 4) << ENA_SHIFT);
+	}
+
+
+	uint32_t ret = en_val + chmux_avl + chmux_val + chmux_sel;
+
+	return ret;
+}
+#endif /* RT 3-digit */
+
+
+static int flexcomm_uart_async_init(const struct device *dev)
+{
+	const struct mcux_flexcomm_config *config = dev->config;
+	struct mcux_flexcomm_data *data = dev->data;
+
+	if (config->rx_dma.dev == NULL ||
+		config->tx_dma.dev == NULL) {
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->rx_dma.dev) ||
+		!device_is_ready(config->tx_dma.dev)) {
+		return -ENODEV;
+	}
+
+	/* Disable DMA requests */
+	USART_EnableTxDMA(config->base, false);
+	USART_EnableRxDMA(config->base, false);
+
+	/* Route DMA requests */
+#if defined(CONFIG_SOC_SERIES_IMX_RT5XX) || defined(CONFIG_SOC_SERIES_IMX_RT6XX)
+	/* RT 3 digit uses input mux to route DMA requests from
+	 * the UART peripheral to a hardware designated DMA channel
+	 */
+	INPUTMUX_Init(INPUTMUX);
+	INPUTMUX_EnableSignal(INPUTMUX,
+		fc_uart_calc_inmux_connection(config->rx_dma.channel,
+					config->rx_dma.base), true);
+	INPUTMUX_EnableSignal(INPUTMUX,
+		fc_uart_calc_inmux_connection(config->tx_dma.channel,
+					config->tx_dma.base), true);
+	INPUTMUX_Deinit(INPUTMUX);
+#endif /* RT5xx and RT6xx */
+
+	/* Init work objects for RX and TX timeouts */
+	k_work_init_delayable(&data->tx_data.timeout_work,
+			config->tx_timeout_func);
+	k_work_init_delayable(&data->rx_data.timeout_work,
+			config->rx_timeout_func);
+
+	return 0;
+}
+
+#endif /* CONFIG_UART_ASYNC_API */
+
+
+static void mcux_flexcomm_isr(const struct device *dev)
+{
+	struct mcux_flexcomm_data *data = dev->data;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	if (data->irq_callback) {
+		data->irq_callback(dev, data->irq_cb_data);
+	}
+#endif
+
+#ifdef CONFIG_UART_ASYNC_API
+	const struct mcux_flexcomm_config *config = dev->config;
+
+	/* If there is an async callback then we are using async api */
+	if (data->async_callback) {
+
+		/* Handle RX interrupt (START bit detected)
+		 * RX interrupt defeats the purpose of UART ASYNC API
+		 * because core is involved for every byte but
+		 * it is included for compatibility of applications.
+		 * There is no other way with flexcomm UART to handle
+		 * Zephyr's RX ASYNC API. However, if not using the RX
+		 * timeout (timeout is forever), then the performance is
+		 * still as might be expected.
+		 */
+		if (config->base->INTSTAT & USART_INTSTAT_START_MASK
+			&& data->async_callback) {
+
```

Comment:
Don't need `&& data->async_callback` as the check is already made earlier.

Reply:
good catch I left those by accident